### ISSUE TITLE
mean-IoU

### ISSUE THREAD

**Issue by rahman-mdatiqur** [qubvel/segmentation_models.pytorch#226]
Hello @qubvel ,

Thank you very much for this really wonderful repo. I just wanted to confirm that the [utils.metrics.IoU](https://github.com/qubvel/segmentation_models.pytorch/blob/master/segmentation_models_pytorch/utils/metrics.py#L6) class is actually computing IoU and not mean-IoU. Please correct me if I am getting it wrong.

Thanks.

**Comment by Hzzone** [comment]
It is only right when the batch_size is set 1.

**Comment by atique81** [comment]
Hello @Hzzone,

thanks for the reply. Can you please explain how that is true even when batch_size = 1? In my understanding, all it is doing is the following - 
`IoU=(Intersection_class1 + Intersection_class2 ......... Intersection_classk) / (Union_class1 + Union_class2  + ........ Union_classK)`

whereas, the mean-IoU measure should be the following - 

`mean-IoU=1/K (Intersection_class1/Union_class1 + Intersection_class2/Union_class2 ......... Intersection_classk/Union_classK)`

**Comment by Hzzone** [comment]
I have not considered the multi-classification you have met, and I agree with the code is only for binary classification, but we should also be known that the following code from this repo only works when batch size is 1:
```python
    intersection = torch.sum(gt * pr)
    union = torch.sum(gt) + torch.sum(pr) - intersection + eps

```
The main reason is that the mean IoU is the mean IoU of different samples, not the intersection/union of all samples.

Maybe the right solution is as follows:
```python
    def iou(self, pr, gt):
        pr = pr > self.threshold
        intersection = torch.sum(pr * gt, dim=[1, 2, 3])
        union = torch.sum(pr, dim=[1, 2, 3]) + torch.sum(gt, dim=[1, 2, 3]) - intersection
        return torch.mean((intersection + self.eps) / (union + self.eps))
```

**Comment by atique81** [comment]
Hello @Hzzone ,
thanks again for your quick and thoughtful feedback. I think, your proposed code is still mixing up the IoU of the different classes as before. I believe, batch size and hence dim=0 is not the issue, rather the issue is the class dimension, dim=1, as we should calculate IoUs for the different classes across the channel dim=1 first, and then take the mean of these IoUs. 

Anyways, I am using a different script for this. Thanks again for your help.

**Comment by JulienMaille** [comment]
@atique81 can you show your loss calculation? I believe the batch also bring issues when averaging IoU of each batch to get the total IoU of an epoch

**Comment by atique81** [comment]
Hello @JulienMaille,

I think, we are talking about the fact that the script [utils.metrics.IoU](https://github.com/qubvel/segmentation_models.pytorch/blob/master/segmentation_models_pytorch/utils/metrics.py#L6) is not computing mean-IoU rather IoU, even for a batch size of 1. So, accumulating IoU (or, mean-IoU) over the whole epoch is not an issue when you have a batch_size of 1. But, that script give different results on a batch_size of 1 when compared with other scripts available online. So, the loss has nothing to do with the issue we are discussing here. But for your info, I am using Cross Entropy loss.

Now, when your batch_size is > 1, then accumulating mean-IoU's (I am not sure about IoUs) across the batches is not what the mean-IoU equation says, rather you have to calculate the intersection and union over all the images in the epoch. This is usually done using a confusion matrix as shown in this [script](https://github.com/davidtvs/PyTorch-ENet/blob/master/metric/iou.py) that keeps a rolling total count of true positives, false positives over all the images in the epoch.

Thanks
